users/models.py

```diff
@@ -1,3 +1,246 @@
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    deleted_at = models.DateTimeField(blank=True, null=True)
+    last_login = models.DateTimeField(null=True, blank=True)
+    original_email = models.EmailField(max_length=255, blank=True, null=True)
+
+    objects = CustomUserManager()
+
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name', 'user_phone']
+
+    def __str__(self):
+        """Return a string representation of the user."""
+        return self.email
+
+    def get_full_name(self):
+        """Return the user's full name."""
+        return f"{self.first_name} {self.last_name}"
+
+    def get_roles_display(self):
+        """Return a string representation of the user's roles."""
+        return ', '.join(name for value, name in self.ROLE_CHOICES if self.role & value)
+
+    def add_role(self, role_name):
+        """
+        Add a role to the user.
+
+        Args:
+            role_name (str): Name of the role to add.
+
+        Raises:
+            ValidationError: If the role is invalid or already assigned.
+        """
+        role_dict = {r[1].upper(): r[0] for r in self.ROLE_CHOICES}
+        if role_name.upper() not in ['INVESTOR', 'STARTUP']:
+            raise ValidationError("You can only add Investor or Startup roles.")
+        if self.has_role(role_name):
+            raise ValidationError(f"You already have the role: {role_name}")
+        self.role |= role_dict[role_name.upper()]
+        self.save()
+
+    def remove_role(self, role_name):
+        """
+        Remove a role from the user.
+
+        Args:
+            role_name (str): Name of the role to remove.
+
+        Raises:
+            ValidationError: If the role is invalid or not assigned.
+        """
+        role_dict = {r[1].upper(): r[0] for r in self.ROLE_CHOICES}
+        if role_name.upper() not in ['INVESTOR', 'STARTUP']:
+            raise ValidationError("Users can only remove Investor or Startup roles.")
+        if not self.has_role(role_name):
+            raise ValidationError(f"User does not have the role: {role_name}")
+        self.role &= ~role_dict[role_name.upper()]
+        if self.role == 0:
+            self.role = self.USER
+        self.save()
+
+    def has_role(self, role_name):
+        """
+        Check if the user has a specific role.
+
+        Args:
+            role_name (str): Name of the role to check.
+
+        Returns:
+            bool: True if the user has the role, False otherwise.
+        """
+        role_dict = {r[1].upper(): r[0] for r in self.ROLE_CHOICES}
+        return self.role & role_dict.get(role_name.upper(), 0) != 0
+
+    def is_admin(self):
+        """Check if the user has admin role."""
+        return bool(self.role & self.ADMIN)
+
+    def has_multiple_roles(self):
+        """Check if the user has multiple roles."""
+        return bin(self.role).count('1') > 1
```

Comment:
In case of a complex role structure, consider changing the role management from bitmasking to many-to-many relations.

Reply:
Agree. The logic of construction is incorrect here